File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/src/DASarrayFn.c
//
//
//  Created by James Ramsay on 2018-3-6.
//
//

#include <stdio.h>

void DASarrayFn(int    *nXbasisw, int    *nWbasisw, int    *nUbasisj, int *nAbasisj,
                int    *nrep,     double *Bvecw,    double *Ucoefj,
                double *BAtens,   double *DASarray)
{
    
    int nXw = *nXbasisw;
    int nWw = *nWbasisw;
    int nUj = *nUbasisj;
    int nAj = *nAbasisj;
    int nRp = *nrep;
    
    int nDASarray = nXw*nRp*nAj;
    
    /* assign 0 to all positions */
    
    for (int p = 0; p < nDASarray; p++)
    {
        DASarray[p]=0;
    }
    
    /* loop through all columns of all four input matrices */
    
    for (int r = 0; r < nRp; r++)
    {
        for (int i = 0; i < nXw; i++)
        {
            for (int j = 0; j < nWw; j++)
            {
                for (int k = 0; k < nUj; k++)
                {
                    for (int l = 0; l < nAj; l++)
                    {
                        int ijkl = i*nAj*nUj*nWw + j*nAj*nUj + k*nAj + l;
                        
                        /*  increment value at index by the product */
                        
                        DASarray[i+r*nXw+l*nXw*nRp] +=
                               Bvecw[j] * Ucoefj[k+r*nUj] * BAtens[ijkl];
                        
                    }
                }
            }
        }
    }
}

File: MSc-SimSci/ACM41000-Uncertainty-Quantification/assignement3/Data2LD_Fix/Data2LD_Fix/R/DASarrayFn.R
DASarrayFn <- function(nXbasisw, nWbasisw, nUbasisj, nAbasisj, 
                      nrep, Bvecw, Ucoefj, BAtens) {
  BAtens <- as.vector(BAtens)
  DSvecC <- rep(0,nXbasisw*nrep*nAbasisj)
  inprodDAList <- .C("DASarrayFn", 
                     as.integer(nXbasisw), 
                     as.integer(nWbasisw), 
                     as.integer(nUbasisj), 
                     as.integer(nAbasisj), 
                     as.integer(nrep),
                     as.double(Bvecw),
                     as.double(Ucoefj),
                     as.double(BAtens), 
                     as.double(DSvecC))
  DASarrayC <- inprodDAList[[9]]
  return(DASarrayC)
}

